nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/spi/PayaraConfig.java

```diff
@@ -177,74 +174,60 @@ private String getValue(String propertyName) {
     }
     
     private <T> Converter<T> getConverter(Class<T> propertyType) {
-        Class type = propertyType;
-        if (type.equals(int.class)) {
-            type = Integer.class;
-        } else if (type.equals(long.class)) {
-            type = Long.class;
-        } else if (type.equals(double.class)) {
-            type = Double.class;
-        } else if (type.equals(float.class)) {
-            type = Float.class;
-        } else if (type.equals(boolean.class)) {
-            type = Boolean.class;
-        }
+        Class type = boxedTypeOf(propertyType);
+
         Converter<T> converter = converters.get(type);
 
+        if (converter != null) {
+            return converter;
+        }
             
-        if (converter == null) {
-            // search for a matching raw type
-            for (Entry<Type, Converter> entry : converters.entrySet()) {
-                Type type1 = entry.getKey();
-                if (type1 instanceof ParameterizedType) {
-                    ParameterizedType ptype = (ParameterizedType)type1;
-                    if (ptype.getRawType().equals(propertyType)) {
-                        converter = entry.getValue();
-                        break;
-                    }
+        // search for a matching raw type
+        for (Entry<Type, Converter> entry : converters.entrySet()) {
+            Type type1 = entry.getKey();
+            if (type1 instanceof ParameterizedType) {
+                ParameterizedType ptype = (ParameterizedType)type1;
+                if (ptype.getRawType().equals(propertyType)) {
+                    return entry.getValue();
```

Comment:
Returning the first converter with matching raw types must be a bug. Say I have a converter `List<String>` and `List<? extends Number>` this could respond with `List<String` when I ask for `List<Integer>` even though there was a proper match.

Reply:
That's right, I think the support for generic types work only for Lists and Sets, and it has specific code in CDI extension for that. I'll remove this part